Cryptodoc Update for ML-KEM (#239)
sehlen-bsi/botan-docs: docs/cryptodoc/src/05_09_kyber.rst

> .. _pubkey/kyber:

Reviewer: ```suggestion
.. _pubkey/ml_dsa:
```

... no strong feeling, but would be more future-oriented.

Author: For sure!

.. _pubkey/kyber/types:

Types
^^^^^

ML-KEM employs strong types and type aliases to represent the various value
types involved in the algorithm. This approach binds the semantic meaning of
values to their types, resulting in a more robust interface and self-documenting
code. Type aliases are defined for ML-KEM polynomials, polynomial vectors, and
polynomial matrices, as well as their NTT representations. All bitstrings,
including various hash values, random seeds, and others, are encapsulated as
strong types. Additionally, the ML-KEM keypair and keypair seed data are
organized within C++ structures.

.. _pubkey/kyber/constants:

Constants
^^^^^^^^^

Botan's ``KyberConstants`` class contains all parameters and constants
outlined in Section 8 of [FIPS-203]_ (see
:ref:`Supported ML-KEM parameter sets <pubkey_key_generation/kyber/table_params>`).
Additionally, the class contains parameters implicitly
derived from these constants, such as key and ciphertext sizes, along with
various intermediate value sizes utilized within the algorithm.

.. _pubkey/kyber/compr_helpers:

Compression Helpers
^^^^^^^^^^^^^^^^^^^

> Botan's helper component implements Formulas 4.7 and 4.8 from [FIPS-203]_

Reviewer 2: *tries to avoid the urge to comment `Formulae` and fails even though Formulas is also correct*

for the compression and decompression of modular ring elements. These operations
involve divisions. While divisions by powers of two can be efficiently executed
in constant time using the right-shift operator, division by the modulus
requires careful handling to avoid timing side-channel leaks, as some compilers
may produce non-constant-time instructions.

To mitigate this risk, Botan employs an alternative division algorithm commonly
used by many compilers for optimization. Based on a technique described in

> [HD]_, this algorithm replaces division with multiplication followed

Reviewer 2: Reference: where is the algorithm / who published this book?

Author: The algorithm should suffice, doesn't it? The complete reference (including author) is in the bibliography.

Reviewer 2: Yeah just saying which algorithm it is suffices imo

Polynomials
^^^^^^^^^^^

ML-KEM relies extensively on polynomials within the polynomial ring :math:`R_q`,
utilizing vectors and matrices of polynomials, both inside and outside the NTT
domain. Botan uses :ref:`strong types <pubkey/kyber/types>` to distinguish
polynomials and polynomial vectors as ``KyberPoly`` and ``KyberPolyVec``, as
well as their NTT counterparts ``KyberPolyNTT`` and ``KyberPolyVecNTT``.
Matrices only appear in the NTT domain and are represented by the class
``KyberPolyMatrix``.

ML-DSA, as defined in [FIPS-204]_, also employs polynomials, leading to shared
polynomial logic between the two algorithms. This shared logic is located in
:srcref:`[src/lib/pubkey]/pqcrystals/pqcrystals.h`, encompassing common
operations on vectors and matrices, as well as algorithm-independent operations
like polynomial addition and subtraction. The ML-KEM specific logic implemented
in :srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber_polynomials.h` supplements
this construction by including the NTT (Algorithm 9 of [FIPS-203]_) and inverse
NTT (Algorithm 10 of [FIPS-203]_) operations, along with NTT polynomial
multiplication (Algorithms 11 and 12 of [FIPS-203]_).

> Due to this type-based construction, the C++ compiler can detect specific

Reviewer 2: ```suggestion
Due to this type-based construction, the C++ compiler can detect domain-specific
```

   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``polynomial_to_message``    | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:212|polynomial_to_message`     | Byte encoding of the K-PKE message :math:`m`                         | 5                         |
   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``expand_keypair``           | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:321|expand_keypair`            | Create public and secret keys from the seeds :math:`d` and :math:`z` | 13, 16                    |
   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+

>    | ``compress_ciphertext``      | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:352|compress_ciphertext`       | Compress, byte encode, and concatenate polynomial vector             | 5, Formula 4.7            |

Reviewer 2: From just reading the document: Why is this now in `kyber_algos` and not, as from the preceding sections, in `kyber_helpers`?

Author: This method does all three things (compression, byte encoding, and concatenation) in one and uses the `kyber_helpers` internally.

   |                              |                                                                                             | :math:`\mathbf{u}` and polynomial :math:`\mathbf{v}`                 |                           |
   +------------------------------+---------------------------------------------------------------------------------------------+----------------------------------------------------------------------+---------------------------+

>    | ``decompress_ciphertext``    | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:362|decompress_ciphertext`     | Split, byte decode, and decompress bytes to polynomial vector        | 6, Formula4.8             |

Reviewer 2: ```suggestion
   | ``decompress_ciphertext``    | :srcref:`[src/lib/pubkey/kyber/kyber_common]/kyber_algos.cpp:362|decompress_ciphertext`     | Split, byte decode, and decompress bytes to polynomial vector        | 6, Formula 4.8             |
```

This class provides an interface for the primitives, which are defined as
:math:`PRF`, :math:`H`, :math:`J`, :math:`G`, and :math:`XOF` in Section 4.1 of
[FIPS-203]_.

.. _pubkey/kyber/kpke_keys:

K-PKE Keys
^^^^^^^^^^

The ``KyberPublicKeyInternal`` and ``KyberPrivateKeyInternal`` classes represent
the public and private keys of ML-KEM, respectively. These classes also provide
methods for K-PKE encryption and decryption, as described in Algorithms 14 and
15 of [FIPS-203]_.

.. _pubkey/kyber/ml_kem_impl:

ML-KEM Implementation
^^^^^^^^^^^^^^^^^^^^^

The ``ML_KEM_Encryptor`` and ``ML_KEM_Decryptor`` classes implement the methods
for high-level and internal ML-KEM encryption and decryption, corresponding to
Algorithms 17, 18, 20, and 21 of [FIPS-203]_.

.. _pubkey/kyber/ml_kem_api:

ML-KEM
^^^^^^

The ``Kyber_PublicKey`` and ``Kyber_PrivateKey`` classes serve as Botan's
public API for public and private ML-KEM keys, respectively. The ``KyberMode``
class is used to select the desired parameter set.

.. _pubkey/kyber/kyber:

Kyber
^^^^^

For compatibility reasons, Botan continues to support the Kyber Round 3.1 NIST

> submission [Kyber-R3]_, which was implemented and released before the release of

Reviewer 2: ```suggestion
[Kyber-R3]_ implementation, which was released before the completion of
```

domain. Botan represents these polynomials and polynomial vectors as
``KyberPoly`` and ``KyberPolyVec``, with their NTT counterparts being
``KyberPolyNTT`` and ``KyberPolyVecNTT``. Matrices only appear in the NTT
domain and are represented by the class ``KyberPolyMatrix``.

ML-DSA, as defined in [FIPS-204]_, also employs polynomials, leading to shared
polynomial logic between the two algorithms. This shared logic is located in
:srcref:`[src/lib/pubkey]/pqcrystals/pqcrystals.h`, encompassing common
operations on vectors and matrices, as well as algorithm-independent operations
like polynomial addition and subtraction. The ML-KEM specific logic implemented
in :srcref:`[src/lib/pubkey/kyber]/kyber_common/kyber_polynomials.h` supplements
this construction by including the NTT (Algorithm 9 of [FIPS-203]_) and inverse
NTT (Algorithm 10 of [FIPS-203]_) operations, along with NTT polynomial
multiplication (Algorithms 11 and 12 of [FIPS-203]_).

Botan's polynomial type system ensures that correct polynomial representations
are used and that operations are valid only on compatible types. For instance,
polynomial multiplication is permissible only if both polynomials are in the

> NTT domain.

Author: Good point

Botan utilizes either Montgomery or Barrett reduction for modular reduction,
depending on the size of the values involved. Montgomery reduction requires
multiplying by a Montgomery factor after each reduction. An optimization from
the reference implementation integrates this factor into the NTT constants,
eliminating the need for an additional multiplication in the NTT
function. For the inverse NTT, another Montgomery factor is embedded to

> compensate for the latest reduction in the main ML-KEM algorithm.

Reviewer: ... also: if someone really wants to go that deep, they'll have to look into the code, and (AFAIK) we did leave a comment about that somewhere. AND, the constexpr-code to derive those NTT-parameters also clearly states that.

In code we trust. 😇

   +------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``compress_ciphertext``      | Compress, byte encode, and concatenate polynomial vector             | 5, Formula 4.7            |
   |                              | :math:`\mathbf{u}` and polynomial :math:`\mathbf{v}`                 |                           |
   +------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``decompress_ciphertext``    | Split, byte decode, and decompress bytes to polynomial vector        | 6, Formula 4.8            |
   |                              | :math:`\mathbf{u'}` and polynomial :math:`\mathbf{v'}`               |                           |
   +------------------------------+----------------------------------------------------------------------+---------------------------+
   | ``sample_matrix``            | Samples a matrix from a secret seed                                  | 7, 13                     |

>    +------------------------------+----------------------------------------------------------------------+---------------------------+

Author: Big thanks for adding these :)

supported methods. The sampling counter :math:`N` is managed by the sampler
object and increments with each method call accordingly.

.. _pubkey/kyber/poly_sample:

.. table::  Polynomial Sampling Methods

   +------------------------------------------+---------------------------------------------------------------------+
   | Polynomial sampler method                | Purpose                                                             |
   +==========================================+=====================================================================+
   | ``sample_polynomial_vector_cbd_eta1``    | Polynomial vector sampling in :math:`\mathcal{D}_{\eta_1}(R_q)`     |
   +------------------------------------------+---------------------------------------------------------------------+
   | ``sample_polynomial_cbd_eta2``           | Polynomial sampling in :math:`\mathcal{D}_{\eta_2}(R_q)`            |
   +------------------------------------------+---------------------------------------------------------------------+
   | ``sample_polynomial_vector_cbd_eta2``    | Polynomial vector sampling in :math:`\mathcal{D}_{\eta_2}(R_q)`     |

>    +------------------------------------------+---------------------------------------------------------------------+

Author: ❤️

^^^^^^^^^^^^^^^^^^^^

In Botan, the symmetric primitives of ML-KEM are represented by the
``KyberSymmetricPrimitives`` class. This class provides an interface for
the primitives, which are defined as :math:`PRF`, :math:`H`, :math:`J`,
:math:`G`, and :math:`XOF` in Section 4.1 of [FIPS-203]_.

.. _pubkey/kyber/kpke_keys:

K-PKE Keys
^^^^^^^^^^

The ``KyberPublicKeyInternal`` and ``KyberPrivateKeyInternal`` classes represent
the public and private keys of ML-KEM, respectively. These classes also provide
methods for K-PKE encryption and decryption, as described in Algorithms 14 and
15 of [FIPS-203]_.

.. _pubkey/kyber/ml_kem_impl:

ML-KEM Implementation
^^^^^^^^^^^^^^^^^^^^^

The ``ML_KEM_Encryptor`` and ``ML_KEM_Decryptor`` classes implement the methods
for high-level and internal ML-KEM encryption and decryption, corresponding to
Algorithms 17, 18, 20, and 21 of [FIPS-203]_.

.. _pubkey/kyber/ml_kem_api:

ML-KEM
^^^^^^

The ``Kyber_PublicKey`` and ``Kyber_PrivateKey`` classes serve as Botan's
public API for public and private ML-KEM keys, respectively. The ``KyberMode``
class is used to select the desired parameter set.

.. _pubkey/kyber/kyber:

Kyber
^^^^^

> For compatibility reasons, Botan continues to support the Kyber Round 3.1 NIST
> submission [Kyber-R3]_. The Kyber and Kyber 90s instances can be activated by
> enabling the ``kyber`` or ``kyber_90s`` module, respectively. This allows for
> the selection of the desired variants within the ``KyberMode``. It is strongly
> recommended to use the ML-KEM instances instead.

Reviewer: Less "technical detail" more "don't touch this".

```suggestion
For compatibility reasons, Botan continues to support the Kyber Round 3.1 NIST
submission [Kyber-R3]_ which was implemented and released before the release of
the final standard. The Kyber and Kyber 90s instances can be activated by
enabling the ``kyber`` or ``kyber_90s`` module, respectively.

Note that the Kyber 90s is already marked as deprecated and both Kyber and Kyber
90s may be removed as early as the next major release of the library. It is not
advisable to use any other variant than the ones specified in [FIPS-203]_.
```

   **Steps:**

   1. Generate the random seed ``seed.d`` and the implicit rejection value ``seed.z`` at random using ``rng``

>    2. ``(rho, sigma) = G(d)``

Reviewer 2: ```suggestion
   2. ``(rho, sigma) = G(d || k)``
```

   3. Sample matrix ``A`` from ``rho`` using ``sample_matrix``
   4. Initialize a ``PolynomialSampler`` ``ps`` with ``sigma``
   5. ``s = ntt(ps.sample_polynomial_vector_cbd_eta1())``
   6. ``e = ntt(ps.sample_polynomial_vector_cbd_eta1())``
   7. Compute ``t = A * s + e``
   8. ``pk = (t, rho)`` and ``sk = (seed.d, seed.z)``

   **Notes:**

   - Step 1 corresponds to Algorithm 19 of [FIPS-203]_ and is performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber.cpp:232|Kyber_PrivateKey::Kyber_PrivateKey`.
   - Steps 2-7 correspond to Algorithms 16 and 13 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_algos.cpp:321|expand_keypair`.

>    - Botan only stores the seeds as the secret key. The required values for decapsulation are recomputed on demand.

Reviewer: ```suggestion
   - Botan only stores the seeds as the secret key. The required values for
     decapsulation are recomputed on demand. Loading or storing the partially
     expanded key format specified in [FIPS-203]_ is explicitly not supported.
```

Author: r

Reviewer 2: Maybe reference the consensus in the LAMPS WG on this? https://mailarchive.ietf.org/arch/msg/spasm/KPg3lbPeNms2H5PNDBt0iJByuE0/

Author: Good point! I'll add a footnote.

.. _pubkey/kyber/encaps:

Key Encapsulation
-----------------

The algorithms for high-level ML-KEM encapsulation and internal encapsulation
(Algorithms 20 and 17 of [FIPS-203]_) are implemented in
:srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:25|ML_KEM_Encryptor::encapsulate`.
They use the K-PKE encapsulation algorithm (Algorithm 14 of [FIPS-203]_)
implemented in
:srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|Kyber_PublicKeyInternal::indcpa_encrypt`.
In combination, Botan does the following:

.. admonition:: ML_KEM_Encryptor::encapsulate

   **Input:**

   -  ``rng``: random number generator
   -  ``pk = (t, rho)``: public key

   **Output:**

   -  ``K``: shared secret key
   -  ``c``: ciphertext

   **Steps:**

   1. Generate a random message ``m`` using ``rng``
   2. ``(K, r) = G(m || H(pk))``
   3. K-PKE encrypt ``m`` using ``r`` to obtain ciphertext ``c``

>       4. Sample transposed matrix ``At`` from ``rho`` using ``sample_matrix``

Reviewer 2: 3.1 etc.

Author: Past-Fabian already did this. Sadly, we cannot write `3.1`. Instead, I begin the indented list with 1 and address the steps in the notes as 3.1., etc.

>       5. Initialize a ``PolynomialSampler`` ``ps`` with ``sigma``

Reviewer 2: ```suggestion
      5. Initialize a ``PolynomialSampler`` ``ps`` with ``r``
```

Author: Good catch! You exposed my copy-paste fingers ;)

      6. ``y = ntt(ps.sample_polynomial_vector_cbd_eta1())``
      7. ``e1 = ps.sample_polynomial_vector_cbd_eta2()``
      8. ``e2 = ps.sample_polynomial_cbd_eta2()``
      9. ``u = inverse_ntt(At * y) + e1``
      10. ``mu = polynomial_from_message(m)`` for byte decoding and decompression
      11. ``v = inverse_ntt(t * y) + e2 + mu``

>       12. Encode and compress ``u`` and ``v`` to obtain ``c = c1 || c2`` using ``compress_ciphertext``
>       13. ``c = c1 || c2``

Reviewer: ```suggestion
      12. Encode, compress and concatenate ``u`` and ``v`` to obtain the
          ciphertext ``c`` using ``compress_ciphertext``
```

... no need to introduce c1 and c2 here, they aren't assigned in the pseudo code beforehand anyway.

Author: c

   **Notes:**

   - Steps 1-3 correspond to Algorithms 20 and 17 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:25|ML_KEM_Encryptor::encapsulate`.

>    - Steps 4-14 correspond to Algorithms 14 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|indcpa_encrypt`.

Reviewer: ```suggestion
   - Steps 4-12 correspond to Algorithm 14 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|indcpa_encrypt`.
```

Author: M

>    - The transposed matrix ``At`` is precomputed and stored in the public key object.

Reviewer: ```suggestion
   - The transposed matrix ``At`` is precomputed and stored in the public key
     object. This way, consecutive encapsulations for the same public key do
     not have to re-generate ``At`` from ``rho``.
```

Author: e

.. _pubkey/kyber/decaps:

Key Decapsulation
-----------------

The algorithms for high-level ML-KEM decapsulation and internal decapsulation
(Algorithms 21 and 18 of [FIPS-203]_) are implemented in
:srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:48|ML_KEM_Decryptor::decapsulate`.

> They uses the K-PKE encapsulation and decapsulation algorithms (Algorithm 14

Reviewer 2: ```suggestion
They use the K-PKE encapsulation and decapsulation algorithms (Algorithm 14
```

and 15 of [FIPS-203]_) implemented in
:srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:55|Kyber_PublicKeyInternal::indcpa_encrypt`
and
:srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:84|Kyber_PrivateKeyInternal::indcpa_decrypt`.
In combination, Botan does the following:

.. admonition:: ML_KEM_Decryptor::decapsulate

   **Input:**

   -  ``c``: ciphertext
   -  ``sk = (seed.d, seed.z)``: secret key
   -  ``pk``: public key

   **Output:**

   -  ``K_prime``: shared secret key

   **Steps:**

>    1. Recompute the secret key value ``s`` from ``seed.d``
>    2. K-PKE decrypt ``c`` to obtain message ``m``

Reviewer: ```suggestion
   1. Recompute the secret key value ``s`` from ``seed.d`` using ``expand_keypair``
   2. K-PKE decrypt ``c`` to obtain message ``m_prime``
```

I think it makes sense to use `m_prime` here, because it is essentially just a decryption "candidate" until the re-encryption was successful. Also FIPS 203 does it like that. Same would be true for u and v, but that would make it harder to read than necessary. Absolutely feel free to omit this suggestion, as it isn't really consistent either!

Author: This was a mistake of me. Even the code says m_prime. Good catch :)

Author: i

>       3. Retrieve ``u, v`` using ``decompress_ciphertext`` on ``c``

Reviewer 2: 2.1 etc.

Author: see above

>       3. Retrieve ``u, v`` using ``decompress_ciphertext`` on ``c``
>       4. Compute ``w = v - inverse_ntt(s * ntt(u))``
>       5. ``m = polynomial_to_message(w)`` for compression and byte encoding

Reviewer: ```suggestion
      3. Retrieve ``u, v`` using ``decompress_ciphertext`` on ``c``
      4. Compute ``w = v - inverse_ntt(s * ntt(u))``
      5. ``m_prime = polynomial_to_message(w)`` for compression and byte encoding
```

>    6. ``(K_prime, r_prime) = G(m || H(pk))``
>    7. ``K_bar = J(seed.z || c)``
>    8. K-PKE encrypt ``m`` using ``r_prime`` to obtain ciphertext ``c_prime``
>    9. if ``c != c_prime`` set ``K_prime = K_bar``

Reviewer: ```suggestion
   6. ``(K_prime, r_prime) = G(m_prime || H(pk))``
   7. ``K_bar = J(seed.z || c)``
   8. K-PKE encrypt ``m_prime`` using ``r_prime`` to obtain ciphertext ``c_prime``
   9. if ``c != c_prime`` set ``K_prime = K_bar``
```

   **Notes:**

   - Steps 1,2 and 6-9 correspond to Algorithm 18 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/ml_kem/ml_kem_impl.cpp:48|ML_KEM_Decryptor::decapsulate`.

>    - Steps 3-5 correspond to Algorithms 15 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:84|Kyber_PrivateKeyInternal::indcpa_decrypt`.

Reviewer: ```suggestion
   - Steps 3-5 correspond to Algorithm 15 of [FIPS-203]_ and are performed in :srcref:`[src/lib/pubkey/kyber]/kyber/kyber_common/kyber_keys.cpp:84|Kyber_PrivateKeyInternal::indcpa_decrypt`.
```
